### ISSUE TITLE
Option for async chunk loading with network error

### ISSUE THREAD

**Issue by 199911** [webpack/webpack#5021]
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

For code splitting with async chunk loading,
when the chunk fail to load, it will call the error callback.
There are no option to have a fallback url for the chunk

**What is the expected behavior?**

By default is load the chunk base on public path, etc `http://cdn.example.com/chunk-1.js`
if `http://cdn.example.com/chunk-1.js` fail to load (404 or timeout),
I can config and fallback path, like `http://another-cdn.example.com`
so that it will try to load `http://another-cdn.example.com/chunk-1.js`

**If this is a feature request, what is motivation or use case for changing the behavior?**

The use case is if the CDN network fail in some region, it can automatically fallback to another CDN or the web server to get the chunk.

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

I am using webpack 2, node v6.9.4, macOS Sierra

**Comment by sokra** [comment]
Try this:

``` js
const cdns = [
	"https://cdn1.example.com/",
	"https://cdn2.example.com/",
	"https://cdn3.example.com/",
];

let currentCdn = 0;

const original = __webpack_chunk_load__;
__webpack_chunk_load__ = (id) => {
	let n = cdns.length;
	return (function tryCdn() {
		if(--n === 0)
			return original(id);
		return original(id).catch(e => {
			currentCdn = (currentCdn + 1) % cdns.length;
			__webpack_public_path__ = cdns[currentCdn];
			return tryCdn();
		})
	}());
}
```

**Comment by 199911** [comment]
@sokra 
Thanks for your quick reply!
But I don't where should i put these code

Do you mean I have to create a plugin?

**Comment by 199911** [comment]
I understand what to do after reading [this](https://webpack.js.org/api/module-variables/)

I have to put the code in my application code and __webpack_chunk_load__ is accessable from the compiled code

**Comment by DmitryGonchar** [comment]
@sokra How to provide fallback if you load only some files from CDN? E.g. angular and jquery which are set `external:` in webpack config (implemented as in the docs), but not the chunks of your own code?

Right now adding `require(file-loader!pathToLibOnOurServer)` throws error `__webpack_public_path__ is not defined`

Or should it be done with the script which is outside of webpack?

**Comment by yss14** [comment]
Is it somehow possible to get chunk information from the chunk id, which is passed to the `__webpack_chunk_load__` function?

**Comment by jedwards1211** [comment]
one thing that confused me at first, which it might help to mention in the docs, is that `__webpack_chunk_load__` gets compiled to `__webpack_require__.e`.

At first I thought it was a local variable and assigning to it would only take effect locally.  It was only when I saw that assigning to it changes the value of `__webpack_require__.e` shared by all modules that I understood why this recommendation works.

**Comment by raronen** [comment]
1) If one uses **MiniCssExtractPlugin** and both chunk.css and chunk.js fail for certain chunkId then there'll be an uncaught exception. It's because they both put a promise on the same installedChunks[chunkId] and the first one that get rejected - free the Promise.all defined in the end of __webpack_require__.e.

To solve this I extended MiniCssExtractPlugin with a custom plugin and in the hook`mainTemplate.hooks.requireEnsure.tap` I replaced `rejecet(err)` with `resolve()`. This way only the chunk.js will fire a fallback to the next cdn.

2) I think we should add this line below, otherwise the first chunk that's failing will force all later loaded chunks to fallback also.
Like this:

> ```js
> const original = __webpack_chunk_load__;
> __webpack_chunk_load__ = (id) => {
> 	 __webpack_public_path__ = cdns[0]     <---------
> 	let n = cdns.length;
> 	return (function tryCdn() {
> 		if(--n === 0)
> 			return original(id);
> 		return original(id).catch(e => {
> 			currentCdn = (currentCdn + 1) % cdns.length;
> 			__webpack_public_path__ = cdns[currentCdn];
> 			return tryCdn();
> 		})
> 	}());
> }
> ```
